File: lab-3-dynamodb/dynamoDBCSharpLab/Lab3/InfectionsDataUploader.cs
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

using Amazon;
using Amazon.DynamoDBv2;
using Amazon.S3;
using Amazon.S3.Model;
using System.Diagnostics;
using System.IO;

namespace Lab3
{
    public static class InfectionsDataUploader
    {
        public static readonly string InfectionsTableName = InfectionsTableCreator.InfectionsTableName;
        public static readonly string S3BucketName = Utils.LabS3BucketName;
        public static readonly string S3BucketRegion = Utils.LabS3BucketRegion;
        public static readonly string InfectionsDataKeyFile = Utils.InfectionsDataKeyFile;

        private static AmazonDynamoDBClient dynamoDBClient = null;
        private static AmazonS3Client s3 = null;

        public static int numberOfFailures = 0;

        public static void Main()
        {
            Init();
            LoadInfectionsData();
        }

        private static void LoadInfectionsData()
        {
            string line;
            char[] splitter = { ',' };
            StreamReader reader = null;

            try
            {
                // Retrieve the infections data file from the S3 bucket
                GetObjectRequest requestFromS3 = null;
                requestFromS3 = new GetObjectRequest()
                {
                    BucketName = S3BucketName,
                    Key = InfectionsDataKeyFile
                };

                using (var responseFromS3 = s3.GetObject(requestFromS3))
                {
                    reader = new StreamReader(responseFromS3.ResponseStream);
                    // Skip first line because it has the attribute names only
                    reader.ReadLine();

                    while ((line = reader.ReadLine()) != null)
                    {
                        // separate CSV values by comma
                        string[] infectionsDataAttrValues = line.Split(splitter);

                        try
                        {
                            if (!infectionsDataAttrValues[0].ToLower().Equals("PatientId"))
                            {
                                // Create an instance of a DynamoDB item with attribute
                                // values read from the infections data file
                                Debug.WriteLine("\n");
                                AddItemToTable(infectionsDataAttrValues);
                                Debug.WriteLine("Added item:" + line);
                            }
                        }
                        catch (AmazonDynamoDBException ex)
                        {
                            Debug.WriteLine("Failed to create item in" + InfectionsTableName);
                            Debug.WriteLine(ex.Message);
                            numberOfFailures++;
                        }
                    }
                }
            }
            catch (FileNotFoundException e)
            {
                Debug.WriteLine(e.StackTrace);
                numberOfFailures = 9999;
            }
            catch (IOException e)
            {
                Debug.WriteLine(e.StackTrace);
                numberOfFailures = 9999;
            }
            finally
            {
                if (reader != null)
                {
                    try
                    {
                        reader.Close();
                    }
                    catch (IOException e)
                    {
                        Debug.WriteLine(e.StackTrace);
                    }
                }
            }

            Debug.WriteLine("Infections data upload complete.");
        }

        private static void Init()
        {
            dynamoDBClient = new AmazonDynamoDBClient();
            RegionEndpoint region = RegionEndpoint.USWest2;
            s3 = new AmazonS3Client(region);
        }

        /**
         * Add the item to the infections data table
         *
         * @param infectionsDataAttrValues    Values to be inserted into the table
         */
        private static void AddItemToTable(string[] infectionsDataAttrValues)
        {
            // STUDENT TODO 3: Replace the solution with your own code
            Solution.AddItemToTable(dynamoDBClient, infectionsDataAttrValues, InfectionsTableName);
        }
    }
}

File: lab-3-dynamodb/dynamoDBJavaLab/src/InfectionsDataUploader.java
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

import com.amazonaws.regions.Region;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.PutItemOutcome;
import com.amazonaws.services.dynamodbv2.document.Table;
import com.amazonaws.services.s3.AmazonS3Client;
import com.amazonaws.services.s3.model.S3Object;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

// The InfectionsDataUploader class reads infections data from a file and uploads each item to the infections table.
public class InfectionsDataUploader {

  public static final String INFECTIONS_TABLE_NAME = InfectionsTableCreator.INFECTIONS_TABLE_NAME;
  public static final Region REGION = Utils.getRegion();
  public static final String S3_BUCKET_NAME = Utils.LAB_S3_BUCKET_NAME;
  public static final String S3_BUCKET_REGION = Utils.LAB_S3_BUCKET_REGION;
  public static final String INFECTIONS_DATA_FILE_KEY = Utils.INFECTIONS_DATA_FILE_KEY;

  private static DynamoDB dynamoDB = null;
  private static AmazonDynamoDBClient dynamoDBClient = null;
  private static AmazonS3Client s3 = null;

  public static int numItemsAdded = 0;

  public static void main(String[] args) throws Exception {

    // Instantiate DynamoDB client and object
    dynamoDBClient = new AmazonDynamoDBClient();
    dynamoDBClient.setRegion(REGION);
    dynamoDB = new DynamoDB(dynamoDBClient);

    // Instantiate S3 client
    s3 = new AmazonS3Client();
    s3.setRegion(Region.getRegion(Regions.fromName(S3_BUCKET_REGION)));

    S3Object infectionsDataObject = null;
    BufferedReader br = null;
    String line = "";
    String splitter = ",";
    PutItemOutcome outcome = null;

    try {
      // Retrieve the infections data file from the S3 bucket
      infectionsDataObject = s3.getObject(S3_BUCKET_NAME, INFECTIONS_DATA_FILE_KEY);
      if (infectionsDataObject == null) {
        System.out.println("Unable to retrieve infections data file");
        return;
      }

      // Retrieve the Table object for the infections table
      Table table = dynamoDB.getTable(INFECTIONS_TABLE_NAME);

      br = new BufferedReader(new InputStreamReader(infectionsDataObject.getObjectContent()));
      // Skip the first line because it contains headings
      br.readLine();

      while ((line = br.readLine()) != null) {
        // Split line into values using comma as the separator
        String[] infectionsDataAttrValues = line.split(splitter);

        if (!infectionsDataAttrValues[0].equals("PatientId")) {

          // Add an item corresponding to the values in the line
          // CSV attributes: PatientId, City, Date
          outcome =
              addItemToTable(
                  table,
                  infectionsDataAttrValues[0],
                  infectionsDataAttrValues[1],
                  infectionsDataAttrValues[2]);

          if (outcome != null) {
            numItemsAdded++;
            System.out.println("Added item:" + line);
          }
        }
      }
    } catch (Exception e) {
      e.printStackTrace();
      numItemsAdded = 0;
    } finally {
      if (br != null) {
        try {
          br.close();
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
      System.out.println("\nNumber of items added: " + numItemsAdded);
    }
    System.out.println("Data upload complete");
  }

  /**
   * Add a record to the DynamoDB table
   *
   * @param table       Table object to update
   * @param patientId   Patient ID
   * @param city        City
   * @param date        Date
   * @return            Addition result
   */
  public static PutItemOutcome addItemToTable(
      Table table, String patientId, String city, String date) {
    // STUDENT TODO 2: Replace the solution with your own code
    return Solution.addItemToTable(table, patientId, city, date);
  }
}
